### ISSUE TITLE
HMM scorer as a complement to the n-gram Markov model

### ISSUE THREAD

**Issue by jonathaneoliver** [jonathaneoliver/infinite-streaming#445]
## Goal

Add an HMM-based session scorer as a research complement to the existing n-gram Markov scorer (`analytics/tools/markov_scorer.py`). The HMM produces a *latent state sequence* per session ("the player was in `RECOVERING` for 18s, then `STEADY`") — information the n-gram model can't surface, which only scores per-transition rarity.

This is a sibling tool, not a replacement. Both run on the same encoded streams and produce parallel reports; a comparison report highlights sessions where they disagree (the most informative cases for understanding what each model captures).

## Design

- **Library**: `hmmlearn` (`CategoricalHMM`)
- **Hidden states**: K=5 default (configurable). Likely emergent: `STARTUP`, `STEADY`, `RECOVERING`, `STALLED`, `ENDING`. Semantic labels assigned post-training by inspecting emission distributions.
- **Observations**: same encoded token alphabet as the Markov scorer (no new tokens)
- **Bucketing**: same `(player_kind, variant_bucket)` cascade as Markov — `MIN_BUCKET_CLEAN_SESSIONS = 5`
- **Initialisation**: random with multiple restarts (5–10), keep best by training likelihood
- **Per-session output**: forward log-likelihood, Viterbi state sequence, RLE state intervals, state-time distribution

## Files

- `analytics/tools/hmm_scorer.py` *(new)* — CLI + training/scoring/report
- `analytics/tools/hmm-scorer.md` *(new)* — README
- `analytics/tools/compare_markov_hmm.py` *(new, small)* — disagreement report
- `analytics/tools/requirements.txt` — add `hmmlearn`
- `analytics/tools/markov_scorer.py` — light refactor of the public-surface helpers so `hmm_scorer.py` can import them; no behaviour change

## Verification

- [ ] Convergence stable across runs on largest bucket (variance <1%)
- [ ] K=5 trained states are labelable from top-5 emissions
- [ ] K-sweep over {3, 5, 7, 9} with BIC; pick lowest BIC for default
- [ ] Spearman ρ between HMM log-likelihood and Markov avg NLL falls in [0.4, 0.7]
- [ ] On a known-bad session, Viterbi state sequence shows meaningful `RECOVERING`/`STALLED` proportion
- [ ] `compare_markov_hmm.md` top-3 disagreements are intuitively explainable
- [ ] 80/20 holdout per bucket — held-out clean log-likelihood overlaps training distribution

## Out of scope

- Porting to Go — both scorers port together when productionization moves forward
- Joint anomaly score combining Markov + HMM — premature
- Semi-supervised HMM — only if v1 unsupervised states are mushy
- Continuous-time HMM — overkill at this stage

Plan in: `~/.claude/plans/telle-me-about-using-memoized-journal.md`

**Comment by jonathaneoliver** [comment]
**Points:** 5 (Fibonacci)
**Effort:** ~1–2 days
**Priority:** P3 — research-direction work; not on the live path. Block behind #442/#443/#444; pick up when Markov scorer is stable and there is appetite for model comparison.
**Value:** medium — latent state sequences are genuinely novel signal vs. the Markov NLL, but the uplift is unproven until verification (K-sweep, Spearman ρ, disagreement spot-check) completes. Upgrade to high if v1 states are clean and the comparison report surfaces actionable anomalies.
**Rationale:** hmmlearn handles Baum-Welch and Viterbi; the work is orchestration + verification. Five points reflects the non-trivial training loop, multi-restart init, K-sweep, and hands-on emission inspection — not algorithm implementation. Calendar effort can stretch to 2 days if the held-out likelihood check reveals bucket-size issues requiring fallback logic.

**Comment by jonathaneoliver** [comment]
## Role clarified (design discussion 2026-05-24)

A design discussion on #442 sharpened where the HMM fits. Capturing it here so this
issue stands on its own.

### The HMM is the home for *regimes*, not separate corpora
A recurring question was whether "normal rate shift" and "normal playback" (or
startup / steady / shifting / rebuffering) should be modelled separately. Answer:
**they're regimes within one session, and the principled way to represent them as
first-class objects is the HMM's hidden states** — not by hand-splitting the training
corpus (which would need phase ground-truth we don't have and would destroy the
boundary transitions where anomalies live).

So the split of labour is:
- **n-gram / variable-order Markov (#442):** "is this *ordering* of requests one we
  see in clean traffic" — structural novelty, scored on tail/peak surprise.
- **HMM (this issue):** latent regime structure — "what mode is the player in, and is
  this regime transition normal" — e.g. rate-shift as a distinct mode from steady
  playback.

### Shared substrate
- Same per-engine + manifest-variant segmentation as #442 (separate grammars), with
  hierarchical back-off across the segmentation tree.
- Same episode-level corpus construction — and the HMM pairs naturally with the
  **contrastive** episode buckets (e.g. shift→stall vs shift→no-stall) discussed on
  #442.
- Same batch labelling / `derived_labels` wiring as #506 (no inline forwarder
  scoring; Python + CH-SQL pushdown; Go only for serving).
- Same make-or-break caveat: only useful where the failure is visible in request
  *structure*; pure timing/throughput failures need timing features, not more model.

See the revised modeling note + runtime addendum on #442 for the full picture.

**Comment by jonathaneoliver** [comment]
### Status refresh — a near-complete draft exists, and the substrate it needs now landed

While cleaning up worktrees I found this issue was **already drafted but never committed** (May 9). I've rescued it from the dangling worktree so it isn't lost:

- **`feat/hmm-scorer-445`** (commit `6b3b8f8`, pushed) — `hmm_scorer.py` (513 lines: `CategoricalHMM` per-bucket training with restarts, Viterbi state decode, BIC, auto-labeled states, RLE intervals, K-sweep, report + CLI), `compare_markov_hmm.py` (disagreement report), `hmm-scorer.md`, and `requirements.txt` (`hmmlearn`, `numpy`, `scikit-learn`).
- **`feat/markov-scorer-442`** (commit `9971b70`, pushed) — the original full #442 `markov_scorer.py` the HMM imports from.

So this is **not greenfield** — there's a working-shaped implementation to revive.

### The catch: it's an older lineage

The draft was built on the **original** `markov_scorer.py` substrate — self-contained, pulls over the **forwarder HTTP API**, with its own tokenizer + player/variant bucketing cascade. That is **not** what got committed. The committed path became the leaner **CH-direct** `scorer.py` (VOMM, PPM-C back-off) + `tokenize.py` in #580. `hmm_scorer.py` imports `TokenVocab` / bucketing / token-encoding from the old `markov_scorer.py`, none of which exist on `dev`.

### Good news: the wiring this issue asked for is now built

The original design notes wanted "same `derived_labels` wiring as #506" — that now exists and has matured:

- **#506** — per-row `derived_labels` (out-of-band batch → derived CH table → read-path LEFT-JOIN merge onto rows + roll-up to play level), surfaced as `unexpected_<condition>` chips (renamed in #594).
- **#608** — split the label deriver into a slow **train** job + a fast **score** job sharing a serialized model artifact, with a deliberately **model-agnostic artifact seam** (`derived_models` manifest + gzipped-JSON artifact on a shared volume). That seam was left generic *for exactly this* — an HMM can publish its own artifact + per-row labels through the same pipeline.

### Reconciliation to revive (the real work now)

1. Re-point `hmm_scorer.py` off the old `markov_scorer.py` helpers onto the current **`tokenize.py`** (same token alphabet, cross-stream episodes) + CH-direct reads (reuse `derive_tokens.py`'s `ch()`/`fetch_rows()`).
2. Emit through the **#608** train/score split rather than its own report-only flow: an `hmm-trainer` (fit + persist artifact) + `hmm-scorer` (Viterbi decode → per-row labels). Latent-state output is a *new label family* (e.g. regime intervals / abnormal regime transitions), distinct from VOMM's point `unexpected_<condition>` surprise.
3. **Packaging cost:** the deriver sidecars run bare `python:3.12-slim` and the tools are currently pure-stdlib. `hmmlearn`/`numpy`/`scikit-learn` means a `requirements.txt` + a pip-install layer (or a purpose-built image) for the HMM sidecar — the main new infra delta vs the stdlib VOMM.

### Why it's still worth it

The split of labour from the earlier discussion holds: VOMM answers "is this *ordering* of requests one we see in clean traffic" (per-transition surprise); the HMM answers "what *regime* is the player in, and is this regime transition normal" (latent-state timeline — `STARTUP`/`STEADY`/`RECOVERING`/`STALLED`/`ENDING`). The comparison report flags sessions where they disagree — the most informative cases. None of that is producible from the point-surprise model we shipped.

Leaving as P3 / open; the draft on `feat/hmm-scorer-445` is the starting point whenever it's picked up.

**Comment by jonathaneoliver** [comment]
## Post-deploy verification (after #612 merges)

The implementation is in #612 (`feat/445-hmm-revive` → `dev`). What it **can't** prove until it runs against live ClickHouse is the #445 verification checklist itself — the synthetic smoke only confirms the plumbing (train → artifact → decode → labels), and its toy-corpus states are deliberately mushy. Run these once the `hmm` profile is deployed:

```bash
# 1. bring up the opt-in sidecars (pip-installs hmmlearn/numpy/scipy on start)
docker compose --profile hmm up -d hmm-label-trainer hmm-label-scorer

# 2. pick K via BIC, then inspect states + separation (local venv against the same CH)
python3 -m venv /tmp/hmm_venv && /tmp/hmm_venv/bin/pip install -r analytics/tools/requirements.txt
/tmp/hmm_venv/bin/python analytics/tools/hmm_scorer.py --days 7 --k-sweep 3,5,7,9 --out /tmp/hmm_sweep
/tmp/hmm_venv/bin/python analytics/tools/hmm_scorer.py --days 7 --k 5 --out /tmp/hmm_report

# 3. where do VOMM and HMM disagree? (pure stdlib, reads derived_labels)
python3 analytics/tools/compare_markov_hmm.py --days 7 --out /tmp/compare.md
```

Checklist to confirm on real data (gates upgrading value:medium → high):

- [ ] K-sweep BIC has a clear minimum → set `HMM_K` (and the deriver `--k`) to it.
- [ ] The K states are labelable from their top emissions (not a wall of `STATE_<i>`).
- [ ] Failed plays' per-token log-likelihood sits clearly below held-out-clean.
- [ ] A known-bad play's ribbon shows meaningful RECOVERING/STALLED time.
- [ ] `compare_markov_hmm.py` Spearman ρ lands in ~[0.4, 0.7], and the top disagreements are intuitively explainable.

If states come out mushy, the documented levers are: bump restarts, widen `--train-days`, or move to per-`(player_kind, variant)` models (the v1 limitation — needs those columns plumbed into the CH fetch first). Full runbook in `analytics/tools/hmm-scorer.md`.

**Comment by jonathaneoliver** [comment]
Done in #612, squash-merged into `dev` (`8dba4c6`). Closing manually since GitHub only auto-closes from PR keywords when merging into the default branch (`main`), not `dev`. Post-deploy verification steps remain tracked in the comment above.
